trunk/src/app/srs_app_edge.cpp

```diff
@@ -840,7 +845,13 @@ srs_error_t SrsEdgeForwarder::cycle()
             return srs_error_wrap(err, "thread pull");
         }
 
-        if ((err = do_cycle()) != srs_success) {
+        // If coroutine stopping, we should always set the quit error code.
+        err = do_cycle();
+        if (send_error_code == 0) {
```

Comment:
```suggestion
      if ((err = do_cycle()) != srs_success) {
            if (send_error_code == ERROR_SUCCESS) {
                // If coroutine stopping, we should always set the quit error code.
                send_error_code = srs_error_code(err);
            }
            return srs_error_wrap(err, "do cycle");
      }
```

suggest another way.

About `send_error_code`, I believe there is a dead loop:
https://github.com/ossrs/srs/blob/8c034e5a13abf209c7cddfb32ae1d32ce6627f09/trunk/src/app/srs_app_edge.cpp#L864-L884

there are no place to set `send_error_code` = ERROR_SUCCESS in the loop, only set the value when `sdk->recv_message(&msg)`.

Reply:
not dead loop, the publish owner will stop it when error.